pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,416 @@
+    df_count_cntrbs = df_in_range.groupby(['Action', 'cntrb_id'])['cntrb_id'].count().to_frame()
+    df_count_cntrbs = df_count_cntrbs.rename(columns={'cntrb_id': 'count'}).reset_index()
+     
+    # pivot df such that the column names correspond to the different action types and the values are the number of contributions of each contributor
+    df_count_cntrbs = df_count_cntrbs.pivot(columns='Action', values='count')
+
+    # initilize empty df with columns that correspond to the different action types to store contributor importance 
+    df_cntrb_importance = pd.DataFrame(columns=df_count_cntrbs.columns)
+    
+    # create a dictionary to store contributor importance for each action type
+    cis = {action_type: 0 for action_type in df_count_cntrbs.columns}
+    for action_type in df_count_cntrbs.columns:
+        # calculate contributor importance for each action time for the specified time window
+        ci = calc_cntrb_importance(df_count_cntrbs, action_type, threshold)
+        cis[action_type] = ci
+    
+    return cis
+
+def process_data(df, patterns, threshold, window_width, step_size, start_date, end_date):
+    # if the start_date and/or the end date is not specified set them to the beginning and most recent created_at date
+    if start_date is None:
+        start_date = df["created_at"].min()
+    if end_date is None:
+        end_date = df["created_at"].max()
+
+    if patterns:
+        # remove rows where Login column value contains the substring 'bot'
+        mask = df['login'].str.contains("|".join(patterns), na=False)
+        df = df[~mask]
+    
+    # convert integer threshold to a float decimal
+    threshold = threshold / 10
+    
+    # convert to datetime objects rather than strings
+    df["created_at"] = pd.to_datetime(df["created_at"], utc=True)
+    start_date = pd.to_datetime(start_date, utc=True)
+    end_date = pd.to_datetime(end_date, utc=True)
+
+    # order values chronologically by COLUMN_TO_SORT_BY date
+    df = df.sort_values(by='created_at', ascending=True)
+    
+    # create bins with a size equivalent to the the step size starting from the start date up to the end date
+    period_from = pd.date_range(start=start_date, end=end_date, freq=f'{step_size}m', inclusive="both")
+    # store the period_from dates in a df 
+    df_time_intervals = period_from.to_frame(index=False, name="period_from")
+    # calculate the end of each interval and store the values in a column named period_from
+    df_time_intervals['period_to'] = df_time_intervals.apply(lambda row: row.period_from + relativedelta(months = window_width), axis=1)
+    # initialize df to store cntrb importance for each action type for each window
+    df_cntrb_importance_over_time = pd.DataFrame(columns = df['Action'].unique())
+
+    # for each starting period, calculate the cntrb importance for each action type
+    for _, row in df_time_intervals.iterrows():
```

Comment:
This can be changed to be done using .apply (much more efficient compared to iterrrows). The logic used in issue_staleness graph can be applied here. Some guidance to get you started (or we can work through this on a call): 
- df_time_intervals is now df_final 
- the function is something like 
`df_final["Commit"], df_final["PR Opened"], df_status["PR Closed"].... (* all other action types here) = zip(
        *df_status.apply(
            lambda row: cntrb_importance_over_time(df, row.period_from, row.period_to, window_width, threshold),
            axis=1,
        )
    )`

-